gnc/sub8_trajectory_generator/include/sub8_trajectory_generator/sub8_state_space.h

```diff
@@ -7,176 +7,218 @@
-			// inherited methods that need to be implemented 
-			//////////////////////////////////////////////////
-
-			// Allocates space for each of the CompoundStateSpace's
-			// subspaces
-			virtual State* allocState() const; 
-
-			// Frees the space allocated to this CompoundStateSpace
-			// and all its subspaces
-			virtual void freeState(State* state) const; 
-
-			// Override the default distance metric, which is to use the L2 norm
-			// Instead, LQR  will be used to calculate a cost in the control space
-			// between two states
-			virtual double distance(const State* state1, const State* state2 ) const; 
-
-			// Get the maximum value a call to distance() can return (or an upper bound).
-			// For unbounded state spaces, this function can return infinity.
-			//
-			// Tight upper bounds are preferred because the value of the extent is used
-			// in the automatic computation of parameters for planning. If the bounds
-			// are less tight, the automatically computed parameters will be less useful.
-			virtual double getMaximumExtent() const; 
-
-			// Define equality between states
-			virtual bool equalStates(const State* state1, const State* state2 ) const; 
-
-		private: 
-			const static int SUB8_STATE_SPACE_ID = 99; 
-		};
-	}
-} 	 	
+namespace trajectory_generator {
+
+// Defines a custom state space, derived from
+// ompl::base::CompoundStateSpace.
+//
+// This state space consists of [q, qdot], where
+// [q] is defined as position and orientation in the world coordinate frame
+// and [qdot] is the time rate of change of [q].
+// See sub8_ode_solver.h for a description of the dynamics of the system
+class Sub8StateSpace : public CompoundStateSpace {
+ public:
+  class StateType : public CompoundStateSpace::StateType {
+   public:
+    StateType() : CompoundStateSpace::StateType() {}
+
+    // The as<SomeStateSpace::StateType>(i) call returns a reference
+    // to the i'th subspace of the CompoundStateSpace, of type "SomeStateSpace"
+
+    /////////////////////////////////////////////////
+    // Getters and Setters, necessary for
+    // sending State obj's to ROS and for
+    // converting Waypoint messages to State obj's
+    /////////////////////////////////////////////////
+
+    // Set x,y,z position
+    void setPosition(double x, double y, double z) const {
+      as<RealVectorStateSpace::StateType>(0)->values[0] = x;
+      as<RealVectorStateSpace::StateType>(0)->values[1] = y;
+      as<RealVectorStateSpace::StateType>(0)->values[2] = z;
+    }
+
+    // Get position as a std::vector<double> pos
+    // pos[0] = x
+    // pos[1] = y
+    // pos[2] = z
+    void getPosition(std::vector<double>& pos) const {
+      if (!pos.empty()) {
+        pos.clear();
+      }
+      pos.push_back(as<RealVectorStateSpace::StateType>(0)->values[0]);
+      pos.push_back(as<RealVectorStateSpace::StateType>(0)->values[1]);
+      pos.push_back(as<RealVectorStateSpace::StateType>(0)->values[2]);
```

Comment:
Yeah, Eigen can be used here. I'm not too familiar with Eigen, and std::vector is perfectly adequate. But I imagine Eigen would make this code cleaner, so will look into refactoring.